drivers/bluetooth/hci/hci_psoc6_bless.c

```diff
@@ -0,0 +1,293 @@
+			if (buf_tailroom < buf_add_len) {
+				LOG_DBG("Not enough space for rx data");
+				return;
+			}
+			net_buf_add_mem(buf, p_hci_rx->data, buf_add_len);
+			bt_recv(buf);
+		} else if (p_hci_rx->packetType == H4_ACL) {
+			buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+			if (!buf) {
+				LOG_ERR("Failed to allocate the buffer for RX: ACL ");
+				return;
+			}
+
+			buf_tailroom = net_buf_tailroom(buf);
+			buf_add_len = p_hci_rx->dataLength;
+			if (buf_tailroom < buf_add_len) {
+				LOG_DBG("Not enough space for rx data");
+				return;
+			}
+			net_buf_add_mem(buf, p_hci_rx->data, p_hci_rx->dataLength);
+			bt_buf_set_type(buf, BT_BUF_ACL_IN);
+			bt_recv(buf);
+
+		} else if (p_hci_rx->packetType == H4_ISO) {
+			if (IS_ENABLED(CONFIG_BT_ISO)) {
+				buf = bt_buf_get_rx(BT_BUF_ISO_IN, K_NO_WAIT);
+				if (!buf) {
+					LOG_ERR("Failed to allocate the buffer for RX: ACL ");
+					return;
+				}
+
+				buf_tailroom = net_buf_tailroom(buf);
+				buf_add_len = p_hci_rx->dataLength;
+				if (buf_tailroom < buf_add_len) {
+					LOG_DBG("Not enough space for rx data");
+					return;
+				}
+				net_buf_add_mem(buf, p_hci_rx->data, p_hci_rx->dataLength);
+				bt_buf_set_type(buf, BT_BUF_ISO_IN);
+				bt_recv(buf);
+			}
+
+		} else {
+			LOG_DBG("Unsupported HCI Package Received");
+			buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+		}
+		break;
+
+	default:
+		LOG_DBG("Other EVENT 0x%X", eventCode);
+		break;
+	}
+}
+
+static void set_public_addr(void)
+{
+	uint8_t *p_addr = (uint8_t *)&SFLASH_BLE_DEVICE_ADDRESS[0];
+	uint8_t addr_buffer[] = {0xA0,	    0xFD,      0x07,	  p_addr[5], p_addr[4],
+				 p_addr[3], p_addr[2], p_addr[1], p_addr[0], 0x00};
+
+	hci_tx_pkt.packetType = 1;
+	hci_tx_pkt.dataLength = 10;
+	hci_tx_pkt.data = &addr_buffer[0];
+
+	Cy_BLE_SoftHciSendAppPkt(&hci_tx_pkt);
+
+	wait_addr_set = true;
+	while (wait_addr_set == true) {
+		Cy_BLE_ProcessEvents();
+		k_msleep(10);
+	}
+}
+
+static int psoc6_bless_hci_init(void)
+{
+	cy_en_ble_api_result_t result;
+	static const struct bt_hci_driver drv = {
+		.name = "PSoC 6 BLESS",
+		.bus = BT_HCI_DRIVER_BUS_VIRTUAL,
+		.quirks = BT_QUIRK_NO_RESET,
+		.open = h4_open,
+		.send = h4_send,
+	};
+
+	/* Connect BLE interrupt to ISR */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), psoc6_bless_isr_handler, 0, 0);
+
+	/* Registers the generic callback functions.  */
+	Cy_BLE_RegisterEventCallback(psoc6_bless_events_handler);
+
+	/* Initializes the PSoC 6 BLESS Controller. */
+	result = Cy_BLE_InitController(&psoc6_bless_config);
+	if (result != CY_BLE_SUCCESS) {
+		LOG_ERR("Failed to init the BLE Controller");
+		return -EIO;
+	}
+
+	/* Enables the BLESS controller in HCI only mode. */
+	result = Cy_BLE_EnableHCIModeController();
+	if (result != CY_BLE_SUCCESS) {
+		LOG_ERR("Failed to enable the BLE Controller in hci mode");
+		return -EIO;
+	}
+
+	/* Enables BLE Low-power mode (LPM)*/
+	Cy_BLE_EnableLowPowerMode();
+
+	/* Set Public address */
+	set_public_addr();
```

Comment:
Seems like the the more correct place to do this would be the optional `setup` callback that you can set in the `bt_hci_driver` struct (you need to also `select BT_HCI_SETUP` in Kconfig). The `setup()` callback will be called after `open()` but before the stack begins the HCI initialization process, and you can use APIs such as `bt_hci_cmd_send_sync()` from there (it seems the public address setting is done using a vendor HCI command). There is currently one upstream driver that takes advantage of this feature, namely `h4.c`, and it delegates the actual sequence to sub-drivers, like here: https://github.com/zephyrproject-rtos/zephyr/blob/2c2d1c72dbbb4b52704e68194a689bf4d573b9df/drivers/bluetooth/hci/cyw43xxx.c#L208